File: 0394-decode-string/0394-decode-string.cpp
class Solution {
public:
    string decodeString(string s) {
        stack<pair<string,int>>st;
        int cn=0;
        string cs="";
        for(char ch:s){
            if(isdigit(ch)){
                cn=cn*10+ch-'0';
            }
         else  if(ch=='['){
                st.push({cs,cn});
                cn=0;
                cs="";
            }
            else if(ch==']'){
                auto[p,n]=st.top();
                st.pop();
                string t="";
                for(int i=0;i<n;i++){
                    t+=cs;
                }
                cs=p+t;
            }
            else{
                cs+=ch;
            }
        }
        return cs;
    }
};

File: 0394-decode-string/0394-decode-string.java
import java.util.Stack;

class Solution {
    public String decodeString(String str) {

        Stack<Integer> n = new Stack<>();
        Stack<StringBuilder> st = new Stack<>();

        StringBuilder ans = new StringBuilder();
        int num = 0;  

        for(char c : str.toCharArray()) {

            if(Character.isDigit(c)) {
                num = (num * 10) + (c - '0');
            }

            else if(c == '[') {
                n.push(num);                           num = 0;

                st.push(ans);         
                ans = new StringBuilder();
            }

            else if(c == ']') {
                String temp = ans.toString();
                ans = new StringBuilder(
                        temp.repeat(n.pop())
                );

                ans = st.pop().append(ans);
            }

            else {
                ans.append(c);
            }
        }

        return ans.toString();
    }
}
